src/cxx_support.rs

```diff
@@ -0,0 +1,214 @@
+pub trait MakeUninitUniquePtr
+where
+  Self: UniquePtrTarget + Sized,
+{
+  // Make a UniquePtr which is allocated (i.e. is not null) but has not yet
+  // had a constructor called.
+  fn make_uninit_unique_ptr() -> MaybeUninitUniquePtr<Self>;
+  // Free a UniquePtr which is allocated (i.e. is not null) but has not yet
+  // had a constructor called.
+  fn free_uninit_unique_ptr(ptr: MaybeUninitUniquePtr<Self>);
+}
+
+impl<T: MakeUninitUniquePtr> Emplace<T> for UniquePtr<T> {
+  fn try_emplace<N: TryNew<Output = T>>(n: N) -> Result<Pin<Self>, N::Error> {
+    unsafe {
+      let mut uninit_ptr = T::make_uninit_unique_ptr();
+      let uninit = uninit_ptr.get_uninitialized_space();
+      // TODO - see if we avoid this transmute.
+      let uninit = std::mem::transmute(uninit);
+      let result = n.try_new(uninit);
+      if let Err(err) = result {
+        T::free_uninit_unique_ptr(uninit_ptr);
+        return Err(err);
+      }
+      // TODO - is it possible to remove the need for Emplace to return a pinned
+      // type?
+      Ok(Pin::new_unchecked(uninit_ptr.assume_init()))
+    }
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use crate::{moveit, Emplace};
+  use cxx::UniquePtr;
+
+  // Shared with C++
+  const UNINITIALIZED: u8 = 0;
+  const INITIALIZED: u8 = 1;
+  const METHOD_CALLED: u8 = 2;
+
+  #[cxx::bridge]
+  pub(super) mod ffi {
+    unsafe extern "C++" {
+      include!("moveit/src/cxx_support_test_cpp.h");
+      type Foo = super::bindgenish::Foo;
+      fn create_uninitialized_foo() -> *mut Foo;
+      unsafe fn free_uninitialized_foo(ptr: *mut Foo);
+
+      fn foo_constructor(_this: Pin<&mut Foo>);
+
+      fn get_status(self: &Foo) -> u8;
+      fn modify(self: Pin<&mut Foo>);
+    }
+    impl UniquePtr<Foo> {}
+  }
+
+  pub(super) mod bindgenish {
+    use std::marker::{PhantomData, PhantomPinned};
+
+    use cxx::{kind::Opaque, type_id, ExternType};
+
+    use crate::{
+      cxx_support::{MakeUninitUniquePtr, MaybeUninitUniquePtr},
+      New,
+    };
+
+    #[repr(C)]
+    pub struct Foo {
+      // opaque
+      _pin: PhantomData<PhantomPinned>,
+      _data: [u32; 4],
+    }
+
+    unsafe impl ExternType for Foo {
+      type Id = type_id!("Foo");
+      type Kind = Opaque;
+    }
+
+    impl MakeUninitUniquePtr for Foo {
+      fn make_uninit_unique_ptr() -> MaybeUninitUniquePtr<Self> {
+        let foo = super::ffi::create_uninitialized_foo();
+        assert_eq!(
+          unsafe { foo.as_ref().unwrap() }.get_status(),
+          super::UNINITIALIZED
+        );
+        unsafe { MaybeUninitUniquePtr::from_ptr(foo) }
+      }
+
+      fn free_uninit_unique_ptr(ptr: MaybeUninitUniquePtr<Self>) {
+        let ptr = unsafe { ptr.get_for_freeing() };
+        unsafe { super::ffi::free_uninitialized_foo(ptr) };
+      }
+    }
+
+    impl Foo {
+      pub(super) fn new() -> impl New<Output = Self> {
+        unsafe {
+          crate::new::by_raw(|space| {
+            // TODO can we get rid of the transmute?
```

Comment:
I don't think so... as I mentioned elsewhere, I think this is the wrong type. You want to be using MaybeUninit<T>.